Hierarchical Bayesian modeling of local adaptation to multidimensional environmental gradients with SNP data.

In development. Please don't hesitate to reach out with ideas or bug reports!

A core question in ecology and evolutionary biology is how multidimensional environmental gradients shape evolutionary adaptation. For example, do genetic variants that increase fitness in dry conditions tend to increase or decrease fitness in cold conditions? The answers hold important implications for conservation strategies like assisted gene flow under climate change. E.g., does importing drought-resistant genotypes increase or decrease population vulnerability to cold snaps?

tradescape is designed for exploring these types of questions with biallelic SNP data--i.e., genomic data for some number of individuals across some number of geographic sampling sites, such that some individuals have one allele at a given point in the genome, and other individuals have a different allele. In this scenario, we can't directly measure how fitness varies for different alleles across different environments, but we can cautiously infer patterns of natural selection through principled modeling of the relationship between environmental conditions and allele frequencies. The core assumption is that, if we've structured our model well, associations between allele frequencies and environmental gradients reflect selection patterns.

A detailed technical treatment of the model structure is on the way. Until then, here is a brief semi-technical summary.

We can think of each SNP as having its own sub-model--effectively, a logit GLMM with a binomial response variable, which is the mean frequency of the reference allele for the focal SNP. (A beta-binomial version is coming soon.) The key terms in the linear predictor model the effects of the environmental variables of interest (e.g., temperature, salinity, etc.) on the reference allele frequency for the focal SNP. Each environmental variable has a linear effect size (on the logit scale) indicating how strongly the mean reference allele frequency increases or decreases with the that variable. Other terms include a SNP-level intercept (these are drawn from a shared distribution across SNPs) and a spatial Gaussian process designed to account for what we might call "isolation by distance" or "neutral demographic history effects".

Now let's zoom out to the level of the whole dataset. This is where we look to answer questions like, "Do alleles that increase with temperature tend to increase or decrease with precipitation?" The key component here is a correlation matrix (derived from a multivariate normal distribution) representing the relationships among the effect sizes for different environmental variables across SNPs. For example, if alleles that increase in frequency with temperature tend to decrease in frequency with precipitation, then we would see a 
